V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-CRYPTO/MASWE-0013.md

Author: Dio: DONE

---
title: Improper Use of Message Authentication Code (MAC)
id: MASWE-0013
alias: improper-mac
requirement: "The app properly uses Message Authentication Codes (MACs)."
platform: [android, ios]
profiles: [L1, L2]
threat: MAS-THREAT-0013

> attacks: [MAS-ATTACK-0018, MAS-ATTACK-0021, MAS-ATTACK-0029]

Reviewer: see attacks.yml
```suggestion
attacks: [MAS-ATTACK-0018, MAS-ATTACK-0021, MAS-ATTACK-0029, MAS-ATTACK-0090]
```

mappings:
  masvs-v1: [MSTG-CRYPTO-4, MSTG-CRYPTO-5]
  masvs-v2: [MASVS-CRYPTO-1, MASVS-CRYPTO-2]

>   cwe: [323, 327, 807, 915]

Reviewer: ```suggestion
  cwe: [208, 323, 327, 354, 807]
```
CWE-915 is "Improperly Controlled Modification of Dynamically-Determined Object Attributes" 
CWE-354 — "Improper Validation of Integrity Check Value"

CWE-208: Observable Timing Discrepancy (see new attack for this one)

- https://developer.android.com/privacy-and-security/cryptography#deprecated-functionality
- https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf
- https://csrc.nist.gov/pubs/sp/800/224/ipd
- https://datatracker.ietf.org/doc/html/rfc6151
- https://web.archive.org/web/20170810051504/http://www.tcs.hut.fi/old/papers/aura/aura-csfws97.pdf
- https://en.wikipedia.org/wiki/Replay_attack
status: new
---

## Overview

This weakness occurs when a Message Authentication Code (MAC) is missing, built on broken primitives, or used incorrectly in a security-sensitive context, so the integrity and authenticity of messages are no longer guaranteed.

A MAC provides integrity and authenticity for a message using a shared secret key. These guarantees only hold when the MAC algorithm is sound, the key is strong and used for a single purpose, the tag is long enough, verification does not leak information, and the protocol prevents captured messages from being accepted again.

## Modes of Introduction

- **Non-Cryptographic Checksums**: Using checksums such as CRC-32 where a MAC is required; checksums detect accidental corruption but can be trivially recomputed by an attacker.

> - **MACs Built on Broken Hashes**: Building MACs on hash functions lacking collision resistance, such as MD5 or SHA-1.

Reviewer: I pulled this interesting [RFC6151](https://datatracker.ietf.org/doc/html/rfc6151):

"attacks on HMAC-MD5 do not seem to indicate a practical vulnerability when used as a message authentication code", but it also adds that "for a new protocol design, a ciphersuite with HMAC-MD5 should not be included"

So I would "losen" this up a bit:
```suggestion
- **MACs Built on Weak Hashes**: Building MACs on deprecated hash functions such as MD5 or SHA-1, or using naive constructions such as `hash(key ‖ message)` that are vulnerable to length-extension attacks.
```
